src/compiler/builder.ts

```diff
@@ -695,12 +695,87 @@ namespace ts {
     export interface ProgramBuildInfo {
         fileInfos: MapLike<BuilderState.FileInfo>;
         options: CompilerOptions;
-        referencedMap?: MapLike<string[]>;
-        exportedModulesMap?: MapLike<string[]>;
+        referencedMap?: MapLike<string[] | number>;
+        referencedMapLists?: string[][];
+        exportedModulesMap?: MapLike<string[] | number>;
+        exportedModulesMapLists?: string[][];

```

Comment:
I think we can simplify and gain even more if we have single list for `referencedMapLists` as well as `exportedModulesMapLists`. Also instead of sometimes deduplicating we can make `referencedMap` and `exportedModulesMap` as `MapLike<number>` and the list of reference set is separate which is deduplicated across the board.

Reply:
I did that, and also deduplicated the files itself.

Sadly that makes all baselines unreadable, maybe we should decompress it for the baseline output to make it readable again...